include/etcpal/cpp/thread.h

```diff
@@ -144,8 +144,11 @@ class Thread
 public:
   /// @brief Create a new thread object which does not yet represent a thread.
   Thread() = default;
-  template <class Function, class... Args>
+  template <class Function,
+            class... Args,
+            typename std::enable_if<!std::is_arithmetic<Function>::value, bool>::type = true>
```

Comment:
What is the purpose of typename here? I thought that keyword was only used for type template parameters, but this appears to be a non-type template parameter. I'm guessing it would compile the same way without the keyword?

Reply:
It actually doesn't compile without `typename`. You'll have to bear with me here, because we are deep enough into template metaprogramming where there is an element of "it's in there because it gets it to compile". But I believe it is necessary because `std::enable_if<Foo>::type` is a [dependent name](https://en.cppreference.com/w/cpp/language/dependent_name), meaning that the compiler does not know whether it names a type or not until it's been specialized. Using `typename` is a way to remove that ambiguity.

The full template argument does resolve to a non-type template parameter, but non-type template parameters still need to start with a type (e.g. bool Foo) so the compiler needs to know that `std::enable_if<!std::is_arithmetic<Function>::value, bool>::type` will indeed be a type name in all cases.

Reply:
In particular, look at the heading "The typename disambiguator for dependent names" on the page linked in the previous comment.

Reply:
Makes perfect sense now, thanks!